File: Algorithm_Solution/baekj/bk_6506.py
# 엘 도라도
import sys

# [A] 입력 함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [B] 메인 로직 함수
def solution(N, K, nums):
    dp = [[0] * (K + 1) for _ in range(N)]
    for i in range(N):
        dp[i][1] = 1
    for i in range(N - 1):
        for j in range(i + 1, N):
            if nums[j] > nums[i]:
                tmp = min(i + 1, K - 1)
                for k in range(1, tmp + 1):
                    dp[j][k+1] += dp[i][k]
    return sum(map(lambda x: x[K], dp))

# [main]
if __name__ == '__main__':
    while True:
        N, K = map(int, input().split())
        if (N, K) == (0, 0): break
        nums = list(map(int, input().split()))
        print(solution(N, K, nums))

File: BOJ_JS/다이나믹 프로그래밍/bk_6506.js
// 엘 도라도

const solution = (input) => {
  let line = 0
  while (true) {
    const [n, k] = input[line++].split(" ").map((v) => +v);
    if (n === 0 && k === 0) break;
    const arr = input[line++].split(" ").map((v) => +v);
    const dp = Array.from({ length: n }, () => 
      Array.from({ length: k + 1 }, () => 0)
    );
    let count = 0;
    for (let i = 0; i < n; i++) {
      dp[i][1] = 1;
      for (let j = 0; j < i; j++) {
        if (arr[j] >= arr[i]) continue;
        let tmp = Math.min(i + 1, k);
        for (let c = 1; c < tmp; c++) {
          dp[i][c + 1] += dp[j][c];
        }
      }
      count += dp[i][k];
    }
    console.log(count);
  }
};

// 데이터 입력 및 메인 함수 실행
const input = require("fs").readFileSync("./input.txt").toString().split("\n");
solution(input);